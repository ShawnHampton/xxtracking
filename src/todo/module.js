import { fromJS } from 'immutable';
import { createSelector } from 'reselect';

export const CONSTANTS = {
	addTodo: 'todo/addTodo',
	addTodoAsync: 'todo/addTodoAsync',
	removeTodo: 'todo/removeTodo'
};

///-- ACTIONS --///

export const addTodo = ({ id, text }) => ({
	type: CONSTANTS.addTodo,
	payload: { id, text }
});
export const removeTodo = id => ({
	type: CONSTANTS.removeTodo,
	payload: { id }
});

const initialState = {};

const reducer = (state = fromJS(initialState), action) => {
	const { type, payload } = action;
	switch (type) {
		case CONSTANTS.addTodo:
		case CONSTANTS.addTodoAsync:
			return state.merge({
				[payload.id]: {
					text: payload.text
				}
			});
		case CONSTANTS.removeTodo:
			return state.delete(payload.id);
		default:
			return state;
	}
};

export default reducer;

///-- SELECTORS --///
const getTodosImmutable = state => state.get('todos', {});
const getTodoImmutable = (state, props) => state.getIn(['todos', props.id], {});

export const getTodoIds = createSelector([getTodosImmutable], todos =>
	todos.keySeq().toArray()
);
export const getTodo = createSelector([getTodoImmutable], todo =>
	todo.toObject()
);
