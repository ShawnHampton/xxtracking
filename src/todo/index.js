import List, { List as ListRaw } from './list';
import Todo, { Todo as TodoRaw } from './todo';
import module from './module';

export { List, ListRaw, Todo, TodoRaw, module };

export default List;
