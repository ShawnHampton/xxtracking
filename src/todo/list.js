import React from 'react';
import Todo from 'todo/todo';
import { connect } from 'react-redux';
import { getTodoIds, addTodo } from 'todo/module';
import { uniqueId } from 'lodash/fp';

export class List extends React.PureComponent {
	state = {
		text: ''
	};
	onAddTodo = () => {
		this.props.addTodo({ id: uniqueId(), text: this.state.text });
	};
	onTextChange = event => this.setState({ text: event.target.value });
	render() {
		const { todoIds } = this.props;
		return (
			<main>
				<h1>Todo List</h1>
				<section>{todoIds.map(id => <Todo key={id} id={id} />)}</section>
				<section>
					<input type="text" onChange={this.onTextChange} />
					<button className="button" onClick={this.onAddTodo}>
						Add Todo
					</button>
				</section>
			</main>
		);
	}
}

const mapStateToProps = state => ({
	todoIds: getTodoIds(state)
});

export default connect(
	mapStateToProps,
	{
		addTodo
	}
)(List);
