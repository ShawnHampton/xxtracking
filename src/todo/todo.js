import React from 'react';
import { connect } from 'react-redux';
import { getTodo, removeTodo } from 'todo/module';

export class Todo extends React.PureComponent {
	onRemove = () => this.props.removeTodo(this.props.id);
	render() {
		const { text } = this.props;
		return (
			<article>
				<span>{text}</span>
				<button onClick={this.onRemove}>Remove</button>
			</article>
		);
	}
}

const mapStateToProps = (state, props) => ({
	text: getTodo(state, props).text
});

export default connect(
	mapStateToProps,
	{
		removeTodo
	}
)(Todo);
