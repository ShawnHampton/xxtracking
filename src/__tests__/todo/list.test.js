import React from 'react';
import { shallow } from 'enzyme';
// Grabbing only the unconnected raw component
import { ListRaw as List } from 'todo';

const config = {
	addTodo: jest.fn(),
	todoIds: ['1', '2']
};

describe('List', () => {
	let wrapper = null;
	beforeEach(() => {
		wrapper = shallow(<List {...config} />);
	});

	describe('Setup and Props', () => {
		test('Match previous snapshot', () => {
			expect(wrapper).toMatchSnapshot();
		});
		test('Match config props', () => {
			expect(wrapper.instance().props).toEqual(expect.objectContaining(config));
		});
	});

	describe('User Input Cases', () => {
		test('On Add Todo', () => {
			wrapper.find('button').simulate('click');
			expect(config.addTodo).toBeCalled();
		});
	});
});
