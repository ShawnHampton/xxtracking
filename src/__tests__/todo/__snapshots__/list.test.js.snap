// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Setup and Props Match previous snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <List
    addTodo={[Function]}
    todoIds={
      Array [
        "1",
        "2",
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1>
          Todo List
        </h1>,
        <section>
          <Connect(Todo)
            id="1"
          />
          <Connect(Todo)
            id="2"
          />
        </section>,
        <section>
          <input
            onChange={[Function]}
            type="text"
          />
          <button
            className="button"
            onClick={[Function]}
          >
            Add Todo
          </button>
        </section>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Todo List",
        },
        "ref": null,
        "rendered": "Todo List",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Connect(Todo)
              id="1"
            />,
            <Connect(Todo)
              id="2"
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "1",
            "nodeType": "class",
            "props": Object {
              "id": "1",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "2",
            "nodeType": "class",
            "props": Object {
              "id": "2",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "section",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              onChange={[Function]}
              type="text"
            />,
            <button
              className="button"
              onClick={[Function]}
            >
              Add Todo
            </button>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "onChange": [Function],
              "type": "text",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Add Todo",
              "className": "button",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Add Todo",
            "type": "button",
          },
        ],
        "type": "section",
      },
    ],
    "type": "main",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1>
            Todo List
          </h1>,
          <section>
            <Connect(Todo)
              id="1"
            />
            <Connect(Todo)
              id="2"
            />
          </section>,
          <section>
            <input
              onChange={[Function]}
              type="text"
            />
            <button
              className="button"
              onClick={[Function]}
            >
              Add Todo
            </button>
          </section>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Todo List",
          },
          "ref": null,
          "rendered": "Todo List",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Connect(Todo)
                id="1"
              />,
              <Connect(Todo)
                id="2"
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "1",
              "nodeType": "class",
              "props": Object {
                "id": "1",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "2",
              "nodeType": "class",
              "props": Object {
                "id": "2",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "section",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                onChange={[Function]}
                type="text"
              />,
              <button
                className="button"
                onClick={[Function]}
              >
                Add Todo
              </button>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "onChange": [Function],
                "type": "text",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Add Todo",
                "className": "button",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Add Todo",
              "type": "button",
            },
          ],
          "type": "section",
        },
      ],
      "type": "main",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
